### PR TITLE
Improve Cron refresh

### PR DESCRIPTION
**Pull request by petems**
Superceeds https://github.com/puppet-community/puppet-mcollective/pull/199

Clean up specs, commits and the like

### CODE REVIEW

File: spec/classes/mcollective_spec.rb [hunk 1]
```diff
@@ -302,6 +303,17 @@ def mcollective_site_libdir_path
...
         end
+
+        describe '#fact_cron_splay' do
+          context 'default (false)' do
+            it { should contain_cron('refresh-mcollective-metadata').with_minute(['0','15','30','45']) }
+          end
+
+          context 'true' do
+            let(:params) { { :fact_cron_splay => true } }
+            it { should contain_cron('refresh-mcollective-metadata').with_minute(anything()) # how do I test randomness? }
```
Comment (petems) [inline spec/classes/mcollective_spec.rb]: Any ideas for this?

**Comment by jyaworski** [comment]
Hello:

What is the status of this PR?

**Comment by petems** [comment]
@jyaworski Right now the code is working, I need help getting the tests to pass. Right now, it fails in travis as the cron results are random (as expected!) but I need to figure out a way to seed the tests so that they give a consitant result so we can test it. 

@nibalizer @igalic any ideas?

**Comment by jyaworski** [comment]
It looks like an off-by-one.

**Comment by jorhett** [comment]
I am trying to find the value provided by this PR. Collecting the facts is trivially easy with near-zero load because it's pulling them from the cached data.

The value of spacing out Puppet runs was to avoid overloading the Puppet server. What value is provided in spacing out fact updates?

**Comment by petems** [comment]
Quoted from the original PR:

refresh-mcollective-metadata is being run at four fixed moments within each hour. In certain configurations with hundreds of virtual machines, all managed by puppet, this simultaneous start of refresh-mcollective-metadata is causing a large CPU steal (ready) time, from seconds to minutes, during which the hypervisors need to deal with the large sudden demand of CPU cycles coming from virtual machines that are agressively competing for CPU resources. To prevent these outages, the refresh-mcollective-metadata cron job should be spaced out over various moments in time, and started less often.

Hypervisor with simultaneous starts of refresh-mcollective-metadata via cron:
image

Hypervisor with spaced out start of the refresh-mcollective-metadata cron jobs (rolling out patch on virtual machines started at 15:20)
image

**Comment by jorhett** [comment]
Re: cpu swarm on hypervisor -- Touché!  I beg forgiveness for having posted that question from a brain rooted in the late 90s by mistake ;)

**Comment by petems** [comment]
No worries! :smile:

I've had people bump into issues like this before, mainly in VMWare clusters, so I wanted to keep the PR alive! :+1:

**Comment by jyaworski** [comment]
@petems is it that the value of fqdn_rand(15, '00:00:00:26:28:8a') is always 1 on that mac and fqdn? You're checking as if the value is zero.

**Comment by petems** [comment]
@jyaworski Good spot, I've fixed it now.

The test failures are now due to https://github.com/rodjek/rspec-puppet/issues/353 and https://github.com/voxpupuli/modulesync_config/pull/65

**Comment by jyaworski** [comment]
Merging despite.

**Comment by neoice** [comment]
this looks like it fixes my issue #207 

when can we expect another Forge release? I really hate pinning to git commits, it can lead to unexpected behavior later.

**Comment by jyaworski** [comment]
@neoice Can you open an issue to issue a forge release? Our process for releasing is very ad-hoc, and an issue is generally the best way to handle it.

**Pull request merged**